### DISCUSSION TITLE
[SOLVED] Gift card work only for staff user despite customers purchase it

### DISCUSSION THREAD

**Discussion by DidoMarchet** [saleor/saleor#9667] [General] [answered]
Hi,
I noticed gift card work only for STAFF USER.

I try to create a gift card to a customer and if he use it Saleor returns:

```
Error: GraphQL error: You need one of the following permissions: OWNER, MANAGE_GIFT_CARD
    at new ApolloError (bundle.esm.js?74ca:63:1)
    at eval (bundle.esm.js?74ca:1247:1)
    at eval (bundle.esm.js?74ca:1559:1)
    at Set.forEach (<anonymous>)
    at eval (bundle.esm.js?74ca:1557:1)
    at Map.forEach (<anonymous>)
    at QueryManager.broadcastQueries (bundle.esm.js?74ca:1555:1)
    at eval (bundle.esm.js?74ca:1646:1)
    at Object.next (Observable.js?a7b2:322:1)
    at notifySubscription (Observable.js?a7b2:135:1)
```

Instead If I use a Staff account it works.

I'm doing something wrong during the creation of the cards?
Or do you have a suggestion?

Because now it's little tricky manage customer with assigned (and purchased) gift card.
 
Thanks in advance and regards,

Davide

**Comment by karolkielecki** [comment, marked as answer]
Hi @DidoMarchet, could you please provide more info? Not sure if I understood correctly. According to our [documentation](https://docs.saleor.io/docs/3.x/developer/gift-cards#creating-gift-cards), only staff users with the MANAGE_GIFT_CARD permission can create gift cards.

**Reply by DidoMarchet** [reply]
Hi @karolkielecki ,
below the steps:

I created gift card and I assigned it to a customer:

<img width="596" alt="Screenshot 2022-04-29 at 14 21 40" src="https://user-images.githubusercontent.com/38940340/165944601-86cb4832-c12d-4349-a68a-6a0d5af2bee6.png">

I get the code:

<img width="617" alt="Screenshot 2022-04-29 at 14 21 47" src="https://user-images.githubusercontent.com/38940340/165944717-adc27853-3488-476a-9a60-7d3a54773be1.png">

In the checkout I log in with the customer account and I use the giftcard:

<img width="1431" alt="Screenshot 2022-04-29 at 14 22 19" src="https://user-images.githubusercontent.com/38940340/165944833-a730f86d-a647-43da-98c5-27cdd07eebfd.png">

And after it return me this error:

<img width="713" alt="Screenshot 2022-04-29 at 14 25 20" src="https://user-images.githubusercontent.com/38940340/165944892-3d1cbfa7-9a52-4bf9-99b0-3d1f5d0438ba.png">


If you need more details let me know,

thanks and regards,

Davide

**Reply by DidoMarchet** [reply]
@karolkielecki 
I tried to buy a giftcard (instead of create and associate it from backend) and it's the same behaviour.

**Reply by karolkielecki** [reply]
Thanks for more details. Maybe it's a bug -- let me check with the team.

**Reply by DidoMarchet** [reply]
Hi @karolkielecki thanks, let me know and if a fix it's possible soon!

Thanks and kind regards,

Davide

**Reply by IKarbowiak** [reply]
Hi @DidoMarchet, could you show what your request payload looks like?

**Reply by DidoMarchet** [reply]
Hi @IKarbowiak!

This is the request I do:

``` graphql
[
   {
      "operationName":"AddCheckoutPromoCode",
      "variables":{
         "checkoutId":"Q2hlY2tvdXQ6NDZmNmE1ODktNDNhMi00NmFiLTk0ZDAtNDNmMDU0ZDk3YWRh",
         "promoCode":"XXXX-XXXX-XXXX"
      },
      "query":"fragment Price on TaxedMoney {\n  gross {\n    amount\n    currency\n    __typename\n  }\n  net {\n    amount\n    currency\n    __typename\n  }\n  __typename\n}\n\nfragment ProductVariant on ProductVariant {\n  id\n  name\n  sku\n  quantityAvailable\n  pricing {\n    onSale\n    priceUndiscounted {\n      ...Price\n      __typename\n    }\n    price {\n      ...Price\n      __typename\n    }\n    __typename\n  }\n  attributes {\n    attribute {\n      id\n      name\n      __typename\n    }\n    values {\n      id\n      name\n      value: name\n      __typename\n    }\n    __typename\n  }\n  product {\n    id\n    name\n    slug\n    thumbnail {\n      url\n      alt\n      __typename\n    }\n    thumbnail2x: thumbnail(size: 510) {\n      url\n      __typename\n    }\n    productType {\n      id\n      isShippingRequired\n      __typename\n    }\n    __typename\n  }\n  __typename\n}\n\nfragment CheckoutLine on CheckoutLine {\n  id\n  quantity\n  totalPrice {\n    ...Price\n    __typename\n  }\n  variant {\n    ...ProductVariant\n    __typename\n  }\n  __typename\n}\n\nfragment Address on Address {\n  id\n  firstName\n  lastName\n  companyName\n  streetAddress1\n  streetAddress2\n  city\n  postalCode\n  country {\n    code\n    country\n    __typename\n  }\n  countryArea\n  phone\n  isDefaultBillingAddress\n  isDefaultShippingAddress\n  __typename\n}\n\nfragment ShippingMethod on ShippingMethod {\n  id\n  name\n  price {\n    currency\n    amount\n    __typename\n  }\n  __typename\n}\n\nfragment PaymentGateway on PaymentGateway {\n  id\n  name\n  config {\n    field\n    value\n    __typename\n  }\n  currencies\n  __typename\n}\n\nfragment Checkout on Checkout {\n  token\n  id\n  totalPrice {\n    ...Price\n    __typename\n  }\n  subtotalPrice {\n    ...Price\n    __typename\n  }\n  billingAddress {\n    ...Address\n    __typename\n  }\n  shippingAddress {\n    ...Address\n    __typename\n  }\n  email\n  availableShippingMethods {\n    ...ShippingMethod\n    __typename\n  }\n  shippingMethod {\n    ...ShippingMethod\n    __typename\n  }\n  shippingPrice {\n    ...Price\n    __typename\n  }\n  lines {\n    ...CheckoutLine\n    __typename\n  }\n  isShippingRequired\n  discount {\n    currency\n    amount\n    __typename\n  }\n  discountName\n  translatedDiscountName\n  voucherCode\n  availablePaymentGateways {\n    ...PaymentGateway\n    __typename\n  }\n  __typename\n}\n\nfragment CheckoutError on CheckoutError {\n  code\n  field\n  message\n  __typename\n}\n\nmutation AddCheckoutPromoCode($checkoutId: ID!, $promoCode: String!) {\n  checkoutAddPromoCode(checkoutId: $checkoutId, promoCode: $promoCode) {\n    checkout {\n      ...Checkout\n      __typename\n    }\n    errors: checkoutErrors {\n      ...CheckoutError\n      __typename\n    }\n    __typename\n  }\n}\n"
   }
]
```

and it returns a checkout object.

After I do:

```graphql
[
   {
      "operationName":"Summary",
      "variables":{
         "token":"46f6a589-43a2-46ab-94d0-43f054d97ada"
      },
      "query":"fragment Price on TaxedMoney {\n  gross {\n    amount\n    currency\n    __typename\n  }\n  net {\n    amount\n    currency\n    __typename\n  }\n  tax {\n    amount\n    currency\n    __typename\n  }\n  __typename\n}\n\nquery Summary($token: UUID!) {\n  checkout(token: $token) {\n    id\n    shippingPrice {\n      ...Price\n      __typename\n    }\n    subtotalPrice {\n      ...Price\n      __typename\n    }\n    discount {\n      amount\n      currency\n      __typename\n    }\n    discountName\n    totalPrice {\n      ...Price\n      __typename\n    }\n    voucherCode\n    giftCards {\n      initialBalance {\n        amount\n        currency\n        __typename\n      }\n      currentBalance {\n        amount\n        currency\n        __typename\n      }\n      displayCode\n      code\n      product {\n        slug\n        __typename\n      }\n      __typename\n    }\n    __typename\n  }\n}\n"
   },
   {
      "operationName":"ShippingMethods",
      "variables":{
         "token":"46f6a589-43a2-46ab-94d0-43f054d97ada"
      },
      "query":"fragment ShippingMethod on ShippingMethod {\n  id\n  name\n  description\n  metadata {\n    key\n    value\n    __typename\n  }\n  price {\n    amount\n    currency\n    __typename\n  }\n  __typename\n}\n\nquery ShippingMethods($token: UUID!) {\n  checkout(token: $token) {\n    id\n    availableShippingMethods {\n      ...ShippingMethod\n      __typename\n    }\n    __typename\n  }\n}\n"
   }
]
```
and it returns:

```graphql
[
   {
      "errors":[
         {
            "message":"You need one of the following permissions: OWNER, MANAGE_GIFT_CARD",
            "locations":[
               {
                  "line":54,
                  "column":7
               }
            ],
            "path":[
               "checkout",
               "giftCards",
               0,
               "code"
            ],
            "extensions":{
               "exception":{
                  "code":"PermissionDenied"
               }
            }
         }
      ],
	  
      /* OTHER DATA CHECKOUT....... */
}
]
```

it the hopes that it is useful
thanks in advance and kind regards,

Davide

**Reply by IKarbowiak** [reply]
So the problem is that you are asking for a gift card `code`. The gift card code can be fetched by the owner of the gift card if it's already used or by staff with `MANAGE_GIFT_CARDS` permission if it isn't. The user is assigned to the gift card when the first order with the given gift card is created. So in this case card is not assigned to anyone so the `code` can be fetched only by staff with the correct permission. 
When you create the gift card from the dashboard you can send it to the specified user, as it was done, but still, the card is not assigned to anyone until the first usage.

**Reply by DidoMarchet** [reply]
Hi @IKarbowiak and thanks.

I did not understand properly.
I created a gift card in the backend and I assigned it to a user.
I bought a gift card from the frontend too.

Both scenarios failed when applying the gift card code.

In the first case I associated a user to the gift card and I tried to use the card even as logged in and not.
It returned the same error.

In the second case I bought the gift card and I tried to use it with the account of the buyer:  it failed too.

Several customers who bought the gift card following the standard checkout flow have the same problem.

We are running:
dashboard v3.1.0-dev core vdev

Thanks in advance for the support,
have a nice day and kind regards!

Davide

**Reply by IKarbowiak** [reply]
You cannot assign a gift card to the user. Until the card is used for the first time, the card is not assigned to anybody. So there is no difference if you are logged in or not before the first order is completed with this gift card. You shouldn't fetch the gift card `code` in the storefront.

**Reply by DidoMarchet** [reply]
Hi @IKarbowiak 
I think I missed something because it return the error when the customer insert the code of the purchased od assigned gift card during the checkout. It returns the error and then the order can't be completed.

Thanks for your patience,

regards,

Davide

**Reply by DidoMarchet** [reply]
Hi @IKarbowiak,
ok I understand what you said with a little delay :D!

I fetch the card because I want to display the residual balance and other info.
But in the query I get the CODE instead of the DisplayCode!

thanks! 
Kind regards again,

Davide

> Hi @IKarbowiak I think I missed something because it return the error when the customer insert the code of the purchased od assigned gift card during the checkout. It returns the error and then the order can't be completed.
> 
> Thanks for your patience,
> 
> regards,
> 
> Davide

**Reply by IKarbowiak** [reply]
I'm glad it's now clear to you! :)

All the bests!
